library/src/scala/deriving/Mirror.scala

```diff
@@ -49,4 +49,15 @@ object Mirror {
   type Of[T] = Mirror { type MirroredType = T; type MirroredMonoType = T ; type MirroredElemTypes <: Tuple }
   type ProductOf[T] = Mirror.Product { type MirroredType = T; type MirroredMonoType = T ; type MirroredElemTypes <: Tuple }
   type SumOf[T] = Mirror.Sum { type MirroredType = T; type MirroredMonoType = T; type MirroredElemTypes <: Tuple }
+
+  extension [T](p: ProductOf[T])
+    /** Create a new instance of type `T` with elements taken from product `a`. */
+    @annotation.experimental
+    def fromProductTyped[A <: scala.Product](a: A)(using m: ProductOf[A], ev: p.MirroredElemTypes =:= m.MirroredElemTypes): T =
```

Comment:
I'm wondering if it is possible to do something like
```scala
    def fromProductTyped[A <: scala.Product](a: A)(using m: ProductOf[A] { type MirroredElemTypes = p.MirroredElemTypes }): T =

```

Reply:
It also feels like `p.MirroredElemTypes =:= m.MirroredElemTypes` should be a `<:<`. As it is safe to construct a `(List[Int], List[String])` from `(::[Int], ::[String])`.

We should have some positive and negative tests for these situations.

Reply:
I didn't use a refinement because of https://github.com/lampepfl/dotty/issues/7380 but just noticed that was fixed. Will give it a shot.